prosody/rootfs/defaults/prosody.cfg.lua

```diff
@@ -215,7 +215,7 @@ http_interfaces = { "*", "::" }
 http_interfaces = { "*" }
 {{ end }}
 
-data_path = "/config/data"
+data_path = "/config/noroot_data"
```

Comment:
Don't do this please. It will break existing installations for no good reason.

Reply:
this is the only way I could get over permission issues while generating TLS certificates by prosody to /config/data.
it appears to me that something sets the permission of /config/data to root no matter what we set it in the Dockerfile. After creating noroot_data and setting the right ownerships, the SSL issue was solved and I was finally able to make a call. otherwise, JVB and Jicofo both failed while trying to reach prosody internally because it doesn't have TLS certificated.

If you have any ideas that can preserve the current setup and still solve the SSL issue with this user I am definitely open to hearing. but I'm curious to know why this will break the existing installations. since the /config/data is still there and the prosody PID is also still in /config/data. this is just changing the data directory to noroot_data to be able to generate SSLs. It appears to me that once the 10-config script is done it will even move the certificates to /config/certs and removes them from noroot_data at the end of the file.

Reply:
What user does the 10-config script run as? A cursory look at https://github.com/bjc/prosody/blob/master/util/prosodyctl/cert.lua reveals some ownership changes if the user is root. Or maybe we should set `prosody_user` and `prosody_group` to the new user on the config file.

Reply:
The user is noroot, with user and group ids = 1000. I bet that's why it can't create the certs in /config/data since apparently we can't change that one's ownership but it can create the certs in noroot_data and move them to the correct location since it's the owner. I could not find a better way to make this work, so I reverted it back to noroot_data so prosody doesn't break.

Reply:
But we can configure the prosody user and group with the options I showed above, did you test that? Why can noroot create stuff in /config/data tough?

We need to fix this before this PR can land.

Reply:
@saghul  I tried the following:

1.  Changing everything back to `/config/data`. 
2. Setting the correct ownership of `noroot:noroot` on `/config` recursively.
3. Setting `prosody_user` and `prosody_group` to 'noroot' in the prosody.cfg.lua as the following: 
  ```
  prosody_user = "noroot"
  prosody_group = "noroot"
```
it still can't write in the /config/data due to the following error:
`The directory /config/data is not owned by the current user, won't be able to write files to it`

Seems like the error is coming from this line of the file you sent before:
https://github.com/bjc/prosody/blob/b5d791c8bffcaa3e200876cb159ff92de2ba1c05/util/prosodyctl/cert.lua#L282

which is specifically checking for the user to be uid 0 (aka root) or own the directory. It is not possible in cloud setups that do not allow running any container as root, so we definitely hit the first part of that `if`. and for the second part, changing the ownership of /config/data does not work either since it reverts back to root:daemon somehow..

Personally, I could not find any solution that works apart from using a backup directory like /config/noroot_data for certs. But I'm open to more suggestions.

Reply:
Since it uses `~=` that line checks if the user is NOT root and is NOT the owner of the certs directory, doesn't it? Maybe try creating the certs directory inside data?

Reply:
@saghul yes precisely. our `uid = 1000` and the owner is `root:daemon`. therefore we hit that if statement and get the error: `The directory `/config/data `is not owned by the current user, won't be able to write files to it`. 

I don't see why putting` /config/certs` directory inside` /config/data `would help solve this problem. I feel like there is a misunderstanding about the nature of the issue. Ownership over /config/certs is **not** the issue here. Ownership over `/config/data` is the actual problem.
Prosody expects noroot to be the owner of that directory. But, regardless of changing that ownership to noroot in the dockerfile, something always overwrites its ownership to `root:daemon`. My guess is it has something to do with `prosody.pid` being there (not actually sure if it can be because of this). 

The bottom line is, if we can't use noroot_data as a temp path for generating certificates, we need to find a way to change ownership of `/config/data`. Otherwise, it will keep trying to write the certs in that one and floods the container log with failed to load TLS certificates errors.

Reply:
> yes precisely. our `uid = 1000` and the owner is `root:daemon`. therefore we hit that if statement and get the error: `The directory `/config/data `is not owned by the current user, won't be able to write files to it`.

Oh hold on. Why can't we change the permissions of that?



> But, regardless of changing that ownership to noroot in the dockerfile, something always overwrites its ownership to `root:daemon`. My guess is it has something to do with `prosody.pid` being there (not actually sure if it can be because of this).

Aha. Maybe try with chattr -i ?

Reply:
@saghul did not work. tried this:
// to install chattr cause it did not exist by default.
RUN apt-get install -y e2fsprogs
// Create /config/data (cause it does not exist at this stage) - change ownership - change attribute and remove immutability.
RUN mkdir -p /config/data && chown -R noroot:noroot /config/data && chattr -R -i /config/data

all of that resulted in:
│ The directory /config/data is not owned by the current user, won't be able to write files to it                            │
│ The directory /config/data is not owned by the current user, won't be able to write files to it                            │
│ mv: cannot stat '/config/data/*.crt': No such file or directory                                                            │
│ mv: cannot stat '/config/data/*.key': No such file or directory

same old same old.

Reply:
Hum, that's very odd, as if the user running `prosodyctl` is root or something. Can you please take care of the rest of the items and leave this one for last? I'll try to take a look too.